File: algorithms/haskell/FibonacciIterative.hs
import Data.Time.Clock
import System.IO
import Control.Exception (evaluate)

modVal :: Integer
modVal = 1000000007

fibIter :: Int -> Integer
fibIter 0 = 0
fibIter n = go n 0 1
  where
    go 1 _ b = b `mod` modVal
    go k a b = go (k-1) b ((a+b) `mod` modVal)

main :: IO ()
main = do
    hSetBuffering stdout LineBuffering
    contents <- getContents
    let n = read (head (words contents)) :: Int

    start  <- getCurrentTime
    result <- evaluate (fibIter n)
    end    <- getCurrentTime

    let ms = realToFrac (diffUTCTime end start) * 1000 :: Double
    print result
    hPutStrLn stderr $ show ms

File: algorithms/lua/fibonacci-iterative.lua
local MOD = 1000000007

local function fib_iterative(n)
    if n == 0 then return 0 end
    local a, b = 0, 1
    for _ = 2, n do
        a, b = b, (a + b) % MOD
    end
    return b
end

local function main()
    local n      = tonumber(io.read("*l"))
    local start  = os.clock()
    local result = fib_iterative(n)
    local ms     = (os.clock() - start) * 1000

    print(result)
    io.stderr:write(string.format("%.3f\n", ms))
end

main()